torch/csrc/jit/passes/graph_fuser.cpp

```diff
@@ -272,15 +261,7 @@ struct GraphFuser {
     // allUsersAreThisConsumerOrOccurAfterIt would still be satisfied. However, I don't expect this
     // to be necessary any time soon, and so we're simply assuming that we don't have to deal with that.
     if (tensors_node->output()->uses().size() > 1) return false;
-    auto tensors = tensors_node->inputs();
-
-    // Our fusion code assumes that all inputs have the same shapes, so we need to check this too.
-    auto expected = tensors.at(0)->type()->cast<TensorType>();
-    if (!expected) return false;
-    return std::all_of(tensors.begin(), tensors.end(), [&expected](Value *v) {
-        auto actual = v->type()->cast<TensorType>();
-        return actual && actual->sizes() == expected->sizes();
-    });
```

Comment:
I am wondering if it is right to delete this code. Another way to do it is to do the test if the information is available, and optimistically assume that things will work out if it is not.

The diff here is a pessimization if it causes you to start fusing more operators which you shouldn't have, because the shapes aren't right. canFuseChunk is particularly dangerous.

Reply:
Err, not quite, right? As in, the old code would simply choose not to consider an operator as part of the fusion group, if it didn't fulfill the necessary conditions for it to be eligible to be in the fusion group.

I can certainly believe in practice that the maximalist fusion always works. But for example, our test suite doesn't have that many "real world" programs in it, does it?

Reply:
Oh, my point is that testing for complete information doesn't really make sense, because it really shouldn't be available to this pass. Optimizations at the GraphExecutor level will see *only* `TensorType` and almost never `CompleteTensorType` (maybe except for some mostly constant subgraphs). 

You do have a point, but:
- It's a temporary state. We need to get this out to unblock splitting the fusers, and the deoptimization path can be easily improved in the future.
- The condition you referred to here is about inputs to concat. I'm pretty sure that the real world examples outside of RNNs will cause us much more trouble because of moving the broadcasts earlier in the graph, or simply having different numbers of elements in different parts of the graph. Irregular cats and chunks are not all that common I think.